Jukebox & Records

JaxkDev opened the pull request:
## Introduction
This PR was mainly so I could ditch my current plugin JukeBox, also gives everyone the relevant items & blocks.
*(Excluding the new `PIGSTEP` record)*

### IMPORTANT NOTE
**Mobile** users **WILL NOT** hear the records being played until they download the `Music` DLC by Mojang in the marketplace.

### Relevant issues
N/A

## Changes
### API changes
<!-- Any additions to the API that should be documented in release notes? -->
Added world sounds:
- RecordSound
- RecordStopSound

Added Blocks:
- Jukebox

Added Items:
- Record
  - 13
  - Cat
  - Blocks
  - Chirp
  - Far
  - Mall
  - Mellohi
  - Stal
  - Strad
  - Ward
  - 11
  - Wait

Added Enums:
- RecordType
  - DISK_13
  - DISK_CAT
  - DISK_BLOCKS
  - DISK_CHIRP
  - DISK_FAR
  - DISK_MALL
  - DISK_MELLOHI
  - DISK_STAL
  - DISK_STRAD
  - DISK_WARD
  - DISK_11
  - DISK_WAIT

Added functions:
 - `Player->sendJukeboxPopup(string key, string[] args)`
 - `NetworkSession->onJukeboxPopup(string key, string[] args)`

### Behavioural changes
N/A

## Backwards compatibility
N/A

## Follow-up

## Tests
[Travis log](https://api.travis-ci.com/v3/job/367679902/log.txt)
All tests completed via travis-ci.com

JaxkDev committed:
Feature - JukeBox/Records.
- Added JukeBox block & tile.
- Added record items excluding new Pigstep record.
- Added record sounds.

JaxkDev committed:
Drop record when jukebox destroyed.

dktapps commented:
You need to run `tests/phpunit/block/regenerate_consistency_check.php` and commit the changes it makes.

JaxkDev committed:
Consistency check regenerated.

JaxkDev commented:
Changes added, all travis tests passed, [new log](https://api.travis-ci.com/v3/job/367641964/log.txt)

dktapps commented on src/block/BlockFactory.php:
```diff
@@ -230,6 +231,7 @@ public function getSilkTouchDrops(Item $item) : array{
 		$this->register(new Opaque(new BID(Ids::IRON_ORE), "Iron Ore", new BlockBreakInfo(3.0, BlockToolType::PICKAXE, ToolTier::STONE()->getHarvestLevel())));
 		$this->register(new Trapdoor(new BID(Ids::IRON_TRAPDOOR), "Iron Trapdoor", new BlockBreakInfo(5.0, BlockToolType::PICKAXE, ToolTier::WOOD()->getHarvestLevel(), 25.0)));
 		$this->register(new ItemFrame(new BID(Ids::FRAME_BLOCK, 0, ItemIds::FRAME, TileItemFrame::class), "Item Frame"));
+		$this->register(new Jukebox(new BID(Ids::JUKEBOX, 0, 0, TileJukebox::class), "Jukebox"));
```
Jukebox item ID is not zero. This will cause bugs when breaking jukeboxes.

dktapps commented on src/block/tile/Jukebox.php:
```diff
@@ -0,0 +1,99 @@
+use pocketmine\world\sound\RecordStopSound;
+
+class Jukebox extends Spawnable{
+	public const TAG_RECORD = "record";
```
This should be `private`. Anything outside of this class has no business using it.
Please also comment next to it what the expected type of tag is.

dktapps replied:
The wiki says that the tag is `RecordItem`. Please check that PC jukeboxes can be decoded correctly.

JaxkDev replied:
Sorry, just copied other tile format most are public with no type.

dktapps commented on src/block/tile/Jukebox.php:
```diff
@@ -0,0 +1,99 @@
+		}
+	}
+
+	protected function addAdditionalSpawnData(CompoundTag $nbt): void{}
```
You're sure about this?

JaxkDev replied:
?

dktapps replied:
It seems unusual that it wouldn't be needed. Doesn't it show the disc in the slot?

JaxkDev replied:
No, unless the particles is client side *shrug*

dktapps commented on src/item/Record.php:
```diff
@@ -0,0 +1,36 @@
+	public function __construct(ItemIdentifier $identifier, string $name)
+	{
+		parent::__construct($identifier, $name);
+	}
```
This constructor seems to be redundant.

JaxkDev replied:
Sorry had them originally pass things but stripped out later.

dktapps commented:
Hi, this looks OK functionally, but it's not quite in line with development practices of PM4.

- Tiles shouldn't do anything more than just handle data storage. Any APIs should be exposed on the corresponding Block class and should give data to the tile in `readStateFromWorld`/`writeStateToWorld` (see for example `ItemFrame`.
- Avoid relying on item IDs or protocol IDs. I recommend creating a `RecordType` enum (like `DyeColor`) and inject the `RecordType` into the item's constructor.

dktapps commented on src/block/tile/Jukebox.php:
```diff
@@ -0,0 +1,99 @@
+	private $record = null;
+
+	public function __construct(World $world, Vector3 $pos){
+		parent::__construct($world, $pos);
```
useless constructor here

JaxkDev committed:
Changes based off review.

JaxkDev commented:
Love you too phpstan :/

dktapps commented on src/block/Jukebox.php:
```diff
@@ -0,0 +1,115 @@
+				$this->ejectRecord();
+			} else {
+				if($item instanceof Record) {
+					$player->getInventory()->remove($item);
```
This will remove **all** record items of this type from the player's inventory. Probably not what you intended.

dktapps commented on src/block/Jukebox.php:
```diff
@@ -0,0 +1,115 @@
+	}
+
+	public function ejectRecord() : void{
+		if($this->hasRecord()){
```
Use `record !== null` here. It's better for static analysis.

dktapps commented on src/block/Jukebox.php:
```diff
@@ -0,0 +1,115 @@
+
+	public function onBreak(Item $item, ?Player $player = null): bool
+	{
+		$this->ejectRecord();
```
It might be better to just add this to the result of `getDrops()` so that plugins can see it in `BlockBreakEvent`.

dktapps reviewed (commented):
This is starting to look better :thumbsup:

JaxkDev committed:
Changes based off review.

JaxkDev committed:
getRecord return type corrected.

JaxkDev committed:
Formatting correction...

dktapps commented on src/block/Jukebox.php:
```diff
@@ -89,10 +89,8 @@ public function stopSound() : void{
-		$this->ejectRecord();
-		return parent::onBreak($item, $player);
+	public function getDrops(Item $item): array{
+		return $this->record !== null ? [$this->record] : [];
```
This breaks the regular jukebox block drop.

dktapps replied:
You also don't want to return the original item, since the caller might modify it.

JaxkDev replied:
I will get there 😊

JaxkDev committed:
Changes based off review.

dktapps commented on src/block/Jukebox.php:
```diff
@@ -90,7 +90,7 @@ public function stopSound() : void{
 	}
 
 	public function getDrops(Item $item): array{
```
This really ought to use `getDropsForCompatibleTool()`, in case a plugin changes the tool requirements.

dktapps commented on src/item/Record.php:
```diff
@@ -0,0 +1,44 @@
+	/** @var RecordType */
+	private $recordType;
+
+	public function __construct(ItemIdentifier $identifier, RecordType $recordType, string $name = "Unknown"){
```
I don't think it makes sense to allow the name to be optional here. It's just asking for bugs.

dktapps commented:
I'll review this again tomorrow.

JaxkDev committed:
Update based off review.

JaxkDev commented:
Reference: LevelSoundEventPacket pigste[ record sound ID is 314

With that i can fully implement the pigstep record to complete all records but its very new and haven't seen any new items or constants is this for a reason ?

dktapps commented:
Avoid introducing any items that were added in 1.13+. World save handling might get real complicated if we introduce those things to an old world format.

dktapps committed:
CS fixes

dktapps committed:
Jukebox: do not make assumptions about parent's getDropsForCompatibleTool()

dktapps committed:
Jukebox: simplify condition for interaction

dktapps committed:
Jukebox is saved as RecordPlayer before PC 1.11

dktapps committed:
Merge remote-tracking branch 'origin/master' into jukebox

dktapps committed:
Jukebox: allow the normal item consuming routine to run its course

dktapps committed:
Jukebox: fixed sound not stopping when breaking block

dktapps committed:
Jukebox: fixed particles not rendering when a record is playing

dktapps commented:
In-game testing revealed the following problems:

- [x] Note particles weren't being displayed while a disk was active. (Missing client-sided blockentity.)
- [x] Records keep playing when the jukebox playing them is destroyed
- [ ] Records are stopped when any jukebox is interacted with or destroyed, regardless of whether the target jukebox is the one currently playing a disc.
- [ ] `Now playing` tip is not displayed
- [x] Disc items were consumed in creative mode
- [ ] Break info is wrong - breaking animation doesn't match when broken by hand (unsure why, haven't investigated).

dktapps commented:
Seems like the hardness is 0.8 on MCPE and not 2.0.

dktapps committed:
Jukebox: hardness is 0.8, not 2.0
it's not clear if this is a bug in MCPE, but it's the same as note block, which makes some sense ...

JaxkDev committed:
Added jukebox popup and the relevant API's and translation keys.

record translations wont work due it being a parameter in the playing now translation...

JaxkDev commented:
Will update PR info to reflect latest commit changes, Cannot use the record name translations for some reason *shrug* i assume its because its a parameter for the jukebox ('playing now') translation

JaxkDev committed:
Record translations used in 'now playing' translation.

dktapps committed:
formatting

dktapps committed:
updated VanillaBlocks and VanillaItems
